crates/stages/src/stages/headers.rs

```diff
@@ -74,63 +74,59 @@ impl<DB: Database, D: HeaderDownloader, C: Consensus, H: HeadersClient, S: Statu
         tx: &mut Transaction<'_, DB>,
         input: ExecInput,
     ) -> Result<ExecOutput, StageError> {
-        let stage_progress = input.stage_progress.unwrap_or_default();
-        self.update_head::<DB>(tx, stage_progress).await?;
+        let current_progress = input.stage_progress.unwrap_or_default();
+        self.update_head::<DB>(tx, current_progress).await?;
 
         // Lookup the head and tip of the sync range
-        let (head, tip) = self.get_head_and_tip(tx, stage_progress).await?;
+        let (head, tip) = self.get_head_and_tip(tx, current_progress).await?;
         debug!(target: "sync::stages::headers", ?tip, head = ?head.hash(), "Commencing sync");
 
-        let mut current_progress = stage_progress;
-        let mut stream =
-            self.downloader.stream(head.clone(), tip).chunks(self.commit_threshold as usize);
-        // The stage relies on the downloader to return the headers
-        // in descending order starting from the tip down to
-        // the local head (latest block in db)
-        while let Some(headers) = stream.next().await {
-            match headers.into_iter().collect::<Result<Vec<_>, _>>() {
-                Ok(res) => {
-                    info!(target: "sync::stages::headers", len = res.len(), "Received headers");
-                    self.metrics.headers_counter.increment(res.len() as u64);
-
-                    // Perform basic response validation
-                    self.validate_header_response(&res)?;
-                    let write_progress =
-                        self.write_headers::<DB>(tx, res).await?.unwrap_or_default();
-                    current_progress = current_progress.max(write_progress);
-                }
-                Err(e) => {
-                    self.metrics.update_headers_error_metrics(&e);
-                    match e {
-                        DownloadError::Timeout => {
-                            warn!(target: "sync::stages::headers", "No response for header request");
-                            return Err(StageError::Recoverable(DownloadError::Timeout.into()))
-                        }
-                        DownloadError::HeaderValidation { hash, error } => {
-                            error!(target: "sync::stages::headers", ?error, ?hash, "Validation error");
-                            return Err(StageError::Validation { block: stage_progress, error })
-                        }
-                        error => {
-                            error!(target: "sync::stages::headers", ?error, "Unexpected error");
-                            return Err(StageError::Recoverable(error.into()))
-                        }
+        // The downloader returns the headers in descending order starting from the tip
+        // down to the local head (latest block in db)
+        let downloaded_headers: Result<Vec<SealedHeader>, DownloadError> =
+            self.downloader.stream(head.clone(), tip).try_collect().await;
```

Comment:
Is this right? Doesn't this mean that we'll take all the headers from the tip to the head and keep them in memory? And that can be the entire chain the first time we run this, which is something we wanted to avoid? Should this stream instead `.take(self.commit_threshold)` headers before collecting?

Reply:
I assumed that we want to write to db from head onwards and update `stage_progress` after every execution. Will change this.